diesel/src/pg/expression/operators.rs

```diff
@@ -118,3 +116,49 @@ where
         out.push_identifier(C::NAME)
     }
 }
+
+#[derive(Debug, Clone, Copy, QueryId, ValidGrouping)]
+#[doc(hidden)]
+pub struct UnionRange<L, R> {
+    pub(crate) left: L,
+    pub(crate) right: R,
+}
+
+impl<L, R> UnionRange<L, R> {
+    pub fn new(left: L, right: R) -> Self {
+        Self {
+            left,
+            right,
+        }
+    }
+}
+
+impl<L, R, ST> crate::expression::Expression for UnionRange<L, R>
+where
+    L: crate::expression::Expression<SqlType = Range<ST>>,
+    R: crate::expression::Expression<SqlType = Range<ST>>,
+    ST: SqlType + TypedExpressionType,
+{
+    type SqlType = ST;
```

Comment:
That's the problematic line that's causing the error.

You specify here that `range_a.union_range(range_b)` return the inner type of the range instead. You want `type SqlType = Range<ST>` instead, which just says that the operator returns a range as well.

That written: You should likely use the `infix_operator!()` macro instead to generate all this code for you. It should be much easier that way.

Reply:
Thanks for the pointer.
My bad! Should have noticed the `__diesel_infix_operator!`.